File: packages/xpute/runtime/src/ipc/ring.test.ts
// @xpute/runtime/ipc/ring.test.ts

import { assert, assertEquals, assertThrows } from "@std/assert";

import { Errno } from "@xpute/core/status/errno.spec.ts";
import { encoder, TreeReader, TreeView } from "@xpute/core/wire/xtp/mod.ts";
import { Ring, ring_bytes, slots_bytes } from "./ring.ts";
import { FrameFlag } from "./frame.ts";

/** A ring of `capacity` slots of `slot` bytes, its descriptors at 8 and its
 * payloads past them, in a buffer of its own. */
function ring(capacity: number, slot: number): Ring {
  const slot_base = 8 + ring_bytes(capacity);
  return Ring.init(new ArrayBuffer(slot_base + slots_bytes(capacity, slot)), 8, capacity, slot_base, slot);
}

Deno.test("ring - messages come out in order across the wrap, and a full ring refuses", () => {
  const r = ring(4, 64);
  for (let round = 0; round < 3; round++) {
    for (let k = 0; k < 4; k++) assertEquals(r.push(round * 4 + k, 7, FrameFlag.ACKREQ, null, -(round * 4 + k)), Errno.OK);
    assertEquals(r.push(99, 0), Errno.EAGAIN, "a full ring refuses");
    for (let k = 0; k < 4; k++) {
      const at = r.peek();
      assert(at >= 0);
      assertEquals([r.tag(at), r.cmd(at), r.flags(at), r.result(at), r.packet(at)], [round * 4 + k, 7, FrameFlag.ACKREQ, -(round * 4 + k), null]);
      r.advance();
    }
    assertEquals(r.peek(), -1);
  }
  assertEquals([r.head, r.tail], [12, 12], "positions only grow");
});

Deno.test("ring - one payload a slot, so the payloads run out with the slots and never before them", () => {
  const r = ring(8, 64);
  const pkt = (n: number) => encoder.encode(new TreeView().branch((b) => b.u32(n).u64(BigInt(n) << 40n)));
  assertEquals(pkt(1).byteLength, 56);

  for (let n = 1; n <= 8; n++) assertEquals(r.push(n, 0x0101, 0, pkt(n)), Errno.OK, "every slot has a payload of its own");
  assertEquals(r.push(9, 0x0101, 0, pkt(9)), Errno.EAGAIN, "and the ring fills before the payloads do");
  assertEquals(r.len, 8);

  for (let n = 1; n <= 8; n++) {
    const at = r.peek();
    const root = new TreeReader(r.packet(at)!).read_branch();
    assertEquals([r.tag(at), root.at(0).get<number>(), root.at(1).get<bigint>()], [n, n, BigInt(n) << 40n]);
    r.advance();
  }
  assertEquals(r.push(9, 0x0101, 0, pkt(9)), Errno.OK, "a slot freed is a payload freed");

  assertEquals(r.push(10, 0x0101, 0, new Uint8Array(72)), Errno.EMSGSIZE);
  assertThrows(
    () => {
      const at = r.entry_at(r.tail);
      r.push(11, 0);
      r.words[at + 3] = 4;
      r.packet(at);
    },
    Error,
    "outside a slot",
  );
});

File: crates/xpute/runtime/src/ipc/ring.test.rs
// xpute-runtime/ipc/ring.test.rs

use super::*;
use crate::ipc::frame::FrameFlag;
use xpute_core::wire::xtp::{encoder, NodeView, TreeEncoderOptions, TreeReader, TreeView};

/// A ring of `capacity` slots of `slot` bytes, its descriptors at 8 and its
/// payloads past them, in a buffer of its own.
fn ring(capacity: u32, slot: u32) -> (Ring, Box<[u8]>) {
    let slot_base = 8 + ring_bytes(capacity);
    let mut buf = vec![0u8; (slot_base + slots_bytes(capacity, slot)) as usize].into_boxed_slice();
    // SAFETY: the buffer outlives the ring — both are returned together and
    // the caller holds them for the test.
    let r = unsafe { Ring::init(buf.as_mut_ptr(), 8, capacity, slot_base, slot) };
    (r, buf)
}

#[test]
fn ring_messages_come_out_in_order_across_the_wrap_and_a_full_ring_refuses() {
    let (r, _buf) = ring(4, 64);
    let ack = FrameFlag::ACKREQ as u32;
    for round in 0..3u32 {
        for k in 0..4 {
            assert_eq!(r.push(round * 4 + k, 7, ack, None, -((round * 4 + k) as i32)), Errno::OK);
        }
        assert_eq!(r.push(99, 0, 0, None, 0), Errno::EAGAIN, "a full ring refuses");
        for k in 0..4 {
            let at = r.peek();
            assert!(at >= 0);
            let at = at as u32;
            assert_eq!((r.tag(at), r.cmd(at), r.flags(at), r.result(at)), (round * 4 + k, 7, ack, -((round * 4 + k) as i32)));
            assert!(r.packet(at).unwrap().is_none());
            r.advance();
        }
        assert_eq!(r.peek(), -1);
    }
    assert_eq!([r.head(), r.tail()], [12, 12], "positions only grow");
}

#[test]
fn ring_a_packet_written_in_place_reads_as_the_one_pushed_and_one_past_a_slot_is_refused_whatever_the_ring_holds() {
    use xpute_core::wire::xtp::PacketWriter;
    let (r, _buf) = ring(8, 64);
    let write = |n: u32| {
        move |out: &mut [u8]| {
            let mut w = PacketWriter::new(out, 2);
            w.u32(Some(n)).u64(Some((n as u64) << 40));
            w.finish()
        }
    };
    let mut t: TreeView<'_> = TreeView::new();
    t.branch(Some(&mut |b| {
        b.u32(Some(5)).u64(Some(5 << 40));
    }));
    let pushed = encoder().encode(&t, TreeEncoderOptions::default()).unwrap();
    assert_eq!(r.push_in_place(1, 2, 0, 0, write(5)), Errno::OK);
    let at = r.peek() as u32;
    assert_eq!(r.packet(at).unwrap().unwrap(), pushed.as_slice());

    // The same packet is refused the same way with a message in the ring and
    // with none: a slot is a slot, so what fits does not depend on what was
    // sent before it.
    let long = "x".repeat(200);
    let too_long = |out: &mut [u8]| {
        let mut w = PacketWriter::new(out, 1);
        w.str(Some(&long));
        w.finish()
    };
    let tail = r.tail();
    assert_eq!(r.push_in_place(1, 2, 0, 0, too_long), Errno::EMSGSIZE);
    assert_eq!(r.tail(), tail, "and nothing written");
    r.advance();
    assert_eq!(r.push_in_place(1, 2, 0, 0, too_long), Errno::EMSGSIZE, "an empty ring answers the same");

    assert_eq!(r.push_in_place(3, 4, 0, -1, write(6)), Errno::OK);
    let at = r.peek() as u32;
    assert_eq!((r.tag(at), r.cmd(at), r.result(at)), (3, 4, -1));
    let reader = TreeReader::new(r.packet(at).unwrap().unwrap()).unwrap().read_branch().unwrap();
    assert_eq!(reader.at(0).unwrap().get_number().unwrap(), 6.0);
}

#[test]
fn ring_one_payload_a_slot_so_the_payloads_run_out_with_the_slots_and_never_before_them() {
    let (r, _buf) = ring(8, 64);
    let pkt = |n: u32| {
        let mut t: TreeView<'_> = TreeView::new();
        t.branch(Some(&mut |b| {
            b.u32(Some(n)).u64(Some((n as u64) << 40));
        }));
        encoder().encode(&t, TreeEncoderOptions::default()).unwrap()
    };
    assert_eq!(pkt(1).len(), 56);

    for n in 1..=8u32 {
        assert_eq!(r.push(n, 0x0101, 0, Some(&pkt(n)), 0), Errno::OK, "every slot has a payload of its own");
    }
    assert_eq!(r.push(9, 0x0101, 0, Some(&pkt(9)), 0), Errno::EAGAIN, "and the ring fills before the payloads do");
    assert_eq!(r.len(), 8);

    for n in 1..=8u32 {
        let at = r.peek() as u32;
        let root = TreeReader::new(r.packet(at).unwrap().unwrap()).unwrap().read_branch().unwrap();
        assert_eq!(
            (r.tag(at), root.at(0).unwrap().get_number().unwrap() as u32, root.at(1).unwrap().get_u64().unwrap()),
            (n, n, (n as u64) << 40)
        );
        r.advance();
    }
    assert_eq!(r.push(9, 0x0101, 0, Some(&pkt(9)), 0), Errno::OK, "a slot freed is a payload freed");

    assert_eq!(r.push(10, 0x0101, 0, Some(&[0u8; 72]), 0), Errno::EMSGSIZE);
    let at = r.entry_at(r.tail());
    r.push(11, 0, 0, None, 0);
    r.words()[(at + FRAME_PACKET) as usize] = 4;
    assert!(r.packet(at).is_err(), "outside a slot");
}
